packages/contracts/contracts/appDefinitions/HighRollerApp.sol

```diff
@@ -0,0 +1,179 @@
+pragma solidity 0.4.25;
+pragma experimental "ABIEncoderV2";
+
+import "../lib/Transfer.sol";
+
+/*
+  This contract is for a dice rolling game.
+  Two players roll two dice each.
+  The winner is the player whose sum of dice outcomes is highest.
+*/
+contract HighRollerApp {
+
+  enum ActionType {
+    COMMIT_TO_HASH,
+    COMMIT_TO_NUM
+  }
+
+  enum Stage {
+    COMMITTING_HASH, 
+    COMMITTING_NUM, 
+    DONE
+  }
+
+  enum Player {
+    FIRST,
+    SECOND
+  }
+
+  struct AppState {
+    address[2] playerAddrs;
+    Stage stage;
+    bytes32 salt;
+    bytes32 commitHash;
+    uint256 commitNum;
+    uint256 hashNum;
+  }
+
+  struct Action {
+    ActionType actionType;
+    uint256 number;
+    bytes32 actionHash;
+  }
+
+  function isStateTerminal(AppState state)
+    public
+    pure
+    returns (bool)
+  {
+    return state.stage == Stage.DONE;
+  }
+
+  function getTurnTaker(AppState state)
+    public
+    pure
+    returns (Player)
+  {
+    if (state.stage == Stage.COMMITTING_NUM) {
+      return Player.SECOND;
+    }
+    return Player.FIRST;
+  }
+
+  function applyAction(AppState state, Action action)
+    public
+    pure
+    returns (bytes)
+  {
+    AppState memory nextState = state;
+    if (action.actionType == ActionType.COMMIT_TO_HASH) {
+      require(state.stage == Stage.COMMITTING_HASH, "Cannot apply COMMIT_TO_HASH on COMMITTING_HASH");
+      nextState.stage = Stage.COMMITTING_NUM;
+
+      nextState.commitHash = action.actionHash;
+    } else if (action.actionType == ActionType.COMMIT_TO_NUM) {
+      require(state.stage == Stage.COMMITTING_NUM, "Cannot apply COMMITTING_NUM on COMMITTING_NUM");
+      nextState.stage = Stage.DONE;
+
+      nextState.commitNum = action.number;
+    } else {
+      revert("Invalid action type");
+    }
+    return abi.encode(nextState);
+  }
+
+  function resolve(AppState state, Transfer.Terms terms)
+    public
+    pure
+    returns (Transfer.Transaction)
+  {
+    bytes32 salt = state.salt;
+    uint256 playerFirstNumber = state.hashNum;
+    uint256 playerSecondNumber = state.commitNum;
+
+    uint256[] memory amounts = new uint256[](2);
+    address[] memory to = new address[](2);
+    to[0] = state.playerAddrs[0];
+    to[1] = state.playerAddrs[1];
+    bytes32 expectedCommitHash  = keccak256(abi.encodePacked(salt, playerFirstNumber));
+    if (expectedCommitHash == state.commitHash) {
+      amounts = getWinningAmounts(playerFirstNumber, playerSecondNumber, terms.limit);
+    } else {
+      amounts[0] = 0;
+      amounts[1] = terms.limit;
+    }
+
+    bytes[] memory data = new bytes[](2);
+
+    return Transfer.Transaction(
+      terms.assetType,
+      terms.token,
+      to,
+      amounts,
+      data
+    );
+  }
+
+  function getWinningAmounts(uint256 num1, uint256 num2, uint256 termsLimit) 
+    public
+    pure
+    returns (uint256[])
+  {
+    uint256[] memory amounts = new uint256[](2);
+    bytes32 finalHash = calculateFinalHash(num1, num2);
```

Comment:
The name "finalHash" is a bit confusing ... maybe just call it `randomNumber` or `randomSalt` or something

Reply:
There should probably be a function `rollDice(bytes32 randomness) returns (uint8 a, uint8 b)`